homeassistant/components/mqtt/number.py

```diff
@@ -0,0 +1,192 @@
+        """Initialize the MQTT Number."""
+        self._config = config
+        self._unique_id = config.get(CONF_UNIQUE_ID)
+        self._sub_state = None
+
+        self._current_number = None
+
+        device_config = config.get(CONF_DEVICE)
+
+        NumberEntity.__init__(self)
+        MqttAttributes.__init__(self, config)
+        MqttAvailability.__init__(self, config)
+        MqttDiscoveryUpdate.__init__(self, discovery_data, self.discovery_update)
+        MqttEntityDeviceInfo.__init__(self, device_config, config_entry)
+
+    async def async_added_to_hass(self):
+        """Subscribe MQTT events."""
+        await super().async_added_to_hass()
+        await self._subscribe_topics()
+
+    async def discovery_update(self, discovery_payload):
+        """Handle updated discovery message."""
+        config = PLATFORM_SCHEMA(discovery_payload)
+        self._config = config
+        await self.attributes_discovery_update(config)
+        await self.availability_discovery_update(config)
+        await self.device_info_discovery_update(config)
+        await self._subscribe_topics()
+        self.async_write_ha_state()
+
+    async def _subscribe_topics(self):
+        """(Re)Subscribe to topics."""
+
+        @callback
+        @log_messages(self.hass, self.entity_id)
+        def message_received(msg):
+            """Handle new MQTT messages."""
+
+            try:
+                if msg.payload.decode("utf-8").isnumeric():
+                    self._current_number = int(msg.payload)
+                else:
+                    self._current_number = float(msg.payload)
+                self.async_write_ha_state()
+            except ValueError:
+                _LOGGER.warning("We received <%s> which is not a Number", msg.payload)
+
+        self._sub_state = await subscription.async_subscribe_topics(
+            self.hass,
+            self._sub_state,
+            {
+                "state_topic": {
+                    "topic": self._config[CONF_TOPIC],
+                    "msg_callback": message_received,
+                    "qos": self._config[CONF_QOS],
+                    "encoding": None,
+                }
+            },
+        )
+
+    async def async_will_remove_from_hass(self):
+        """Unsubscribe when removed."""
+        self._sub_state = await subscription.async_unsubscribe_topics(
+            self.hass, self._sub_state
+        )
+        await MqttAttributes.async_will_remove_from_hass(self)
+        await MqttAvailability.async_will_remove_from_hass(self)
+        await MqttDiscoveryUpdate.async_will_remove_from_hass(self)
+
+    @property
+    def value(self):
+        """Return the current value."""
+        return self._current_number
+
+    async def async_set_value(self, value: float) -> None:
+        """Update the current value."""
+        if value.is_integer():
+            self._current_number = int(value)
+        else:
+            self._current_number = value
```

Comment:
Can this be covered by tests as well?

Reply:
It's covered in `test_run_number_setup` with "10" and "20.5"